nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/internal/memory/ArrayCacheMemoryMgr.java

```diff
@@ -160,15 +191,17 @@ private static boolean isCpu() {
         return !"CUDA".equalsIgnoreCase(backend);
     }
 
+
+
     @Override
-    public INDArray allocate(boolean detached, DataType dataType, long... shape) {
+    public synchronized INDArray allocate(boolean detached, DataType dataType, long... shape) {
         String arrayShapeString = Arrays.toString(shape);
-        if (arrays.contains(dataType, arrayShapeString) && enableCache) {
+        if (getArraysForThread().contains(dataType, arrayShapeString) && enableCache) {
             INDArray arr = null;
             boolean arrFound = false;
             while(!arrFound) {
-                arr = !arrays.get(dataType, arrayShapeString).isEmpty()
-                        ? arrays.get(dataType, arrayShapeString).remove(0)
+                arr = !getArraysForThread().get(dataType, arrayShapeString).isEmpty()
+                        ? getArraysForThread().get(dataType, arrayShapeString).remove(0)
```

Comment:
Why the repeated calls to the method? The reference shouldn't be changing in between calls, so why not get it in the beginning?